plugins/vpc-shared-eni/network/bridge_windows.go

```diff
@@ -266,8 +278,84 @@ func (nb *BridgeBuilder) FindOrCreateEndpoint(nw *Network, ep *Endpoint) error {
 	return nil
 }
 
+// findOrCreateEndpointNS creates a new HNS endpoint in the namespace.
+func (nb *BridgeBuilder) findOrCreateEndpointNS(nw *Network, ep *Endpoint) error {
+	// Check if the endpoint already exists.
+	endpointName := nb.generateHNSEndpointName(ep, ep.NetNSName)
+	hnsEndpoint, err := hcsshim.GetHNSEndpointByName(endpointName)
+	if err == nil {
+		log.Infof("Found existing HNS endpoint %s, adding to ns: %s", endpointName, ep.NetNSName)
+		err = nb.addNamespaceEndpoint(hnsEndpoint, ep.NetNSName)
+
+		ep.MACAddress, _ = net.ParseMAC(hnsEndpoint.MacAddress)
+		return err
+	}
+
+	// Initialize the HNS endpoint.
+	hnsEndpoint = &hcsshim.HNSEndpoint{
+		Name:               endpointName,
+		VirtualNetworkName: nb.generateHNSNetworkName(nw),
+		DNSServerList:      strings.Join(nw.DNSServers, ","),
+		GatewayAddress:     nw.GatewayIPAddress.String(),
+		IPAddress:          ep.IPAddress.IP,
+	}
+
+	// Attach policies
+	err = nb.addEndpointPolicy(
+		hnsEndpoint,
+		hnsRoutePolicy{
+			Policy:            hcsshim.Policy{Type: hcsshim.Route},
+			DestinationPrefix: ep.IPAddress.IP.String() + "/32",
+			NeedEncap:         true,
+		})
+	if err != nil {
+		log.Errorf("Failed to add endpoint route policy: %v.", err)
+		return err
+	}
+
+	err = nb.addEndpointPolicy(
+		hnsEndpoint,
+		hnsRoutePolicy{
+			Policy: hcsshim.Policy{Type: hcsshim.OutboundNat},
+		})
+	if err != nil {
+		log.Errorf("Failed to add OutboundNat policy: %v.", err)
+		return err
+	}
+
+	hnsResponse, err := hnsEndpoint.Create()
+	if err != nil {
+		log.Errorf("Failed to create HNS endpoint: %v.", err)
+		return err
+	}
+
+	log.Infof("Received HNS endpoint response: %+v.", hnsResponse)
+
+	// Move the HNS endpoint to the namespace.
+	err = nb.addNamespaceEndpoint(hnsResponse, ep.NetNSName)
+	if err != nil {
+		// Cleanup the failed endpoint.
+		log.Infof("Deleting the failed HNS endpoint %s.", hnsResponse.Id)
+		_, delErr := hcsshim.HNSEndpointRequest("DELETE", hnsResponse.Id, "")
+		if delErr != nil {
+			log.Errorf("Failed to delete HNS endpoint: %v.", delErr)
```

Comment:
What would happen if the DEL call fails as well? Will the endpoint be leaked in that case